### PR TITLE
webhooks: Support including only a set of headers

### PR DESCRIPTION
**Pull request by ParkMyCar**
This PR adds support for mapping header values to columns, and optionally filtering the headers in the `headers` column. Specifically:

```
CREATE SOURCE webhook_text_with_selective_headers IN CLUSTER webhook_cluster FROM WEBHOOK
  BODY FORMAT TEXT
  INCLUDE HEADER 'content-type' AS content_type
  INCLUDE HEADER 'x-hash' as hash BYTES
  INCLUDE HEADERS (
    NOT 'x-api-key',
    NOT 'x-authorization',
    'timestamp'
  );
```

The source `webhook_text_with_selective_headers` would then have the following columns:

```
SHOW COLUMNS FROM webhook_text_with_selective_headers;
name         nullable  type
----------------------------
body         false     text
headers      false     map
content_type true      text
hash         true      bytea
```

### Motivation

* This PR adds a feature that has not yet been specified.

As part of the [original design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/20230615_webhook_source.md#future) for things we'd want to do in the future, we described the ability to filter/include only a specific set of headers. This PR implements that feature.

It also fixes a paper cut in supporting https://github.com/MaterializeInc/materialize/issues/21332. AWS EventBridge supports three kinds of validation:

1. Basic Authentication (Username and Password) ❌
2. OAuth2 ❌
3. API Key in a header ✅ 

We cannot support 1 or 2, but we can already support 3. The papercut is that the API key gets included in the headers of the request, and currently you can only include _all headers_ or _no headers_ in your source. So if your event contains a header that you need, the only way you can include that header requires you to also leak your API Key in plain text in your source.

After this PR users can selectively include headers, fixing the issue.

### Tips for reviewer

This PR is broken up into three logical components which can be reviewed independently:

1. Changes to the SQL Parser to support the new syntax
2. Updates to planning and our HTTP handlers to support the specific headers
3. Testdrive and sqllogictests

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Allows users to specify a list of headers to include for webhook sources, instead of including all or none.

### CODE REVIEW

File: src/environmentd/src/http/webhook.rs [hunk 1]
```diff
@@ -124,12 +126,30 @@ fn pack_row(
...
     // Pack the headers into our row, if required.
-    if header_ty.is_some() {
-        packer.push_dict(
-            headers
-                .iter()
-                .map(|(name, val)| (name.as_str(), Datum::String(val))),
-        );
+    match header_tys {
+        Some(include_headers) if include_headers.selective_headers.is_empty() => {
+            packer.push_dict(
```
Comment (mjibson) [inline src/environmentd/src/http/webhook.rs:131]: It is real odd that all headers get included as a dict, whereas named headers as a string. Can we choose one and always use it? Otherwise the validation methods have to differ. Users switching between them will get confused and it'll make the docs weirder.

**Comment by chuck-alt-delete** [comment]
Would it be possible to also have `EXCLUDE HEADERS` in case there are many headers and only a few that a user wants to exclude?

File: src/environmentd/src/http/webhook.rs (hunk 1 above)
Comment (ParkMyCar) [reply src/environmentd/src/http/webhook.rs]: Chatted in [Slack](https://materializeinc.slack.com/archives/C02FWJ94HME/p1693334482449209).

tl;dr to map a header to a column name we'll support `INCLUDE HEADER <header-name> AS <column-name>`. To include all headers as a map and possibly filter them we support `INCLUDE HEADERS ( [NOT] <header-name> )`

**Comment by ParkMyCar** [comment]
> Would it be possible to also have `EXCLUDE HEADERS` in case there are many headers and only a few that a user wants to exclude?

Thanks for the suggestion @chuck-alt-delete! The plan right now is to use a `NOT` keyword to support excluding headers, e.g.
```
INCLUDE HEADERS (
  NOT 'api-key',
  NOT 'authorization
)
```

What do you think?

**Comment by chuck-alt-delete** [comment]
I think that works! It’s a little scraggly when you have some with NOT and others without NOT. I’m assuming that would be equivalent to omitting the NOT ones? I have no objections, but I’m not any kind of authority on syntax

**Review by mjibson — APPROVED** [review]

File: src/sql-parser/src/parser.rs [hunk 2]
```diff
@@ -2848,7 +2848,38 @@ impl<'a> Parser<'a> {
...
             _ => unreachable!(),
         };
-        let include_headers = self.parse_keywords(&[INCLUDE, HEADERS]);
+
+        let mut include_headers = CreateWebhookSourceIncludeHeaders::default();
+        while self.parse_keyword(INCLUDE) {
+            match self.expect_one_of_keywords(&[HEADER, HEADERS])? {
+                HEADER => {
+                    let header_name = self.parse_literal_string()?;
+                    self.expect_keyword(AS)?;
```
[the remark this answers is not in the document]
Comment (benesch) [reply src/sql-parser/src/parser.rs:2857]: That's a bit of a recipe for frustration as header names almost always contain `-` characters which are not valid bare identifiers. Maybe we should have a rule that the default name for a header is the header name with all dashes replaced with underscores? `content-encoding` becomes `content_encoding`, for example.

[the remark this answers is not in the document]
Comment (benesch) [reply src/sql-parser/src/parser.rs:2857]: Also header names can start with a digit apparently, and valid SQL identifiers cannot!

File: src/sql-parser/src/ast/defs/statement.rs [hunk 3]
```diff
@@ -831,6 +829,79 @@ impl AstDisplay for CreateWebhookSourceBody {
...
+    pub header_name: String,
+    pub column_name: Ident,
+    pub use_bytes: bool,
+}
+
+impl AstDisplay for CreateWebhookSourceMapHeader {
+    fn fmt<W: fmt::Write>(&self, f: &mut AstFormatter<W>) {
+        f.write_str("INCLUDE HEADER ");
+
+        f.write_node(&display::escaped_string_literal(&self.header_name));
```
[the remark this answers is not in the document]
Comment (ParkMyCar) [reply src/sql-parser/src/ast/defs/statement.rs:892]: No unfortunately `display::escape_single_quote_string` just escapes the string, it doesn't wrap it in single quotes to make it a string literal

File: src/sql-parser/src/parser.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ParkMyCar) [reply src/sql-parser/src/parser.rs:2857]: Given the possible complexity because of the constraints on what a valid SQL identifier is, I'd prefer to do this as a follow up? Relaxing the requirement for `AS <column-name>` should be backwards compatible change

[the remark this answers is not in the document]
Comment (benesch) [reply src/sql-parser/src/parser.rs:2857]: I think that's a good call.
